# By Fernando Perez
#In this script there are three functions required by the ASCAT3_QCs_ovaHRDscar_onlyTumor.R script
#Each function has description on int

### Function to check if files exists ####
check_file_exists <- function(file_path) {
  if (!file.exists(file_path)) {
    stop(paste("The file does not exist:", file_path))
  } else {
    cat("The file exists:", file_path, "\n")
  }
}


# Script to select only autosomes and order by position the Log2Ratio and BAF files generated by GenomeStudio
# The resultant files will be used as input for ASCAT
sort_select_autosomes <- function(input_file, output_file, input.type="Log.R.Ratio"){
    
    #Checking if input file has the right type designation
    if(input.type == "Log.R.Ratio" | input.type == "B.Allele.Freq"){
      print(paste0("Selected input type: ", input.type))
    }else{
      print(paste0("Not recognized the next input type:", input.type))
      stop('Invalid input type. Valid input.type are: Log.R.Ratio  or B.Allele.Freq')
    }
  
    #Checking if input file exist
    if (!file.exists(input_file)) {
      stop(paste("The file does not exist:", input_file))
    } else {
      cat("The starting to read the file:", input_file, "\n")
    }
  
    print("Validating if input file header has the right format")
    
    # Read header separately
    header <- readLines(input_file, n = 1)
    
    head.col <- read.csv(input_file,  sep="\t", nrows=1L)
    head.col <- colnames(head.col)
    
    #Checking if header have column names in the right order
    if (!(head.col[1] == "Name" && head.col[2] == "Chr" && head.col[3] == "Position")){
      stop("The first three columns of your input file does not have the right names.
           The right name of these columns must be: Name Chr Position")
    }
    
    #Checking if fourth column has the right name
    if (!grepl(input.type, head.col[4])){
      stop(paste0("The fourth columnd most contain:",  input.type, " in the header"))
    }
    
    
    print("Header has the right format")
    n.samples <- length(head.col) - 3
    samples <- head.col[4:length(head.col)]
    print(paste0("Analyzing in total ", n.samples, " samples"))
    print(paste0("This is the list of samples:"))
    print(samples)
    
    print("Reading input file")
    data <- read.delim(input_file, skip = 1, header = FALSE, stringsAsFactors = FALSE)
    

    print("Sorting SNPs based on chromosome position")
    # Filter out rows that contain 'chr' in any column
    data <- data[!apply(data, 1, function(row) any(grepl("chr", row))), ]
    
    print("Selecting autosomes (1-22) and the X chromosome")
    # Separate autosomes (1-22) and chrX
    is_autosome <- grepl("^[1-9]$|^1[0-9]$|^2[0-2]$", data[, 2])
    is_chrX <- data[, 2] == "X"
    
    # Combine autosomes and chrX
    filtered_data <- rbind(
      data[is_autosome, ],
      data[is_chrX, ]
    )
    
    # Sort by chromosome (as number or X) and then position (column 3)
    filtered_data$ChrNumeric <- as.numeric(ifelse(filtered_data[, 2] == "X", 23, filtered_data[, 2]))
    filtered_data <- filtered_data[order(filtered_data$ChrNumeric, as.numeric(filtered_data[, 3])), ]
    
    # Drop the helper column
    filtered_data$ChrNumeric <- NULL
    
    # Write output
    writeLines(header, paste0(output_file, ".build"))
    write.table(filtered_data, file = paste0(output_file, ".build"), sep = "\t",
                row.names = FALSE, col.names = FALSE, quote = FALSE, append = TRUE)
    
    # Rename .build file to final output
    file.rename(paste0(output_file, ".build"), output_file)
    print(paste0("Your file was successfully generated: ", output_file))
}



##Function to calculate paired differences of LogR between consecutive SNPs
#data, is a dataframe with LogR values per probe (SNP), each row is a probe, each column a sample
#For data use the LogR file generate by GenomeStudio
#columns, refers to the columns names
#if you dont use columns names then use columns numbers as input and give that to the the function
#By default GenomeStudio uses column names
get.MAPD <- function(data, columns){
  rows <- list(2:nrow(data))
  sapply(columns, function (x){
    PairDifferences <- sapply(rows, function (y){
      diff <- data[y,x] - data[c(y-1),x]
      return(diff)
    })
    PairDifferences <- PairDifferences[!is.na(PairDifferences)] #Remove NAs
    MAPD <- round(median(abs(PairDifferences)), 3)
    return(MAPD)
  })
}
