# SNParray_ovaHRDscar_QC
Script for calculating ovaHRDscar from LogRatio files and BAF files produced with GenomeStudio

## Getting started

### Requirements

- R 4.2.3
- R packages:
  - usethis
  - devtools
  - dplyr
  - ggrepel
  - ovaHRDscar (https://github.com/farkkilab/ovaHRDscar)

In addition you would need ASCAT v3.1.2. This version is stored in our datacloud: https://datacloud.helsinki.fi/index.php/f/54077565
To install this ASCAT version do the next:
1. Download the ascat-master.zip file
2. Decompress the file
3. In case you have another version of ASCAT already installed, uninstall that ASCAT version you have.
4. Install this ASCAT version from R studio in the next way, just replace YOURPATH for your own computer path where you decompressed the ascat-master.zip file.

```
install_local("YOURPATH/ascat-master/ASCAT/", force = TRUE)
```

### Input files

**1. LogR ratio files and BAF files from GenomeStudio:**

From Illumina-SNP arrays raw data the LogR ratio and BAF files can be generated using the Genome Studio software ([link](https://www.illumina.com/techniques/microarrays/array-data-analysis-experimental-design/genomestudio.html)).

Here there is a tutorial about how to prepare a Genome Studio project: <https://www.youtube.com/embed/s23379Gya0Y?autoplay=1&rel=0>

Once that the project is setup and the call rates have been generated, next select the columns necessary for this script and export those in a text file. Here is a tutorial that describe how to select the columns: [link](http://penncnv.openbioinformatics.org/en/latest/user-guide/input/#preparing-input-signal-intensity-files-from-beadstudio-project-files)

Also here there is information about how  to generate the input files: https://docs.google.com/document/d/1mXT1G87VnemVMd0mZsD9MbP4VigF6vtQ/edit

Example of input LogRratio file:

```
Name	Chr	Position	206064570174_R01C01.Log R Ratio	206064570174_R01C02.Log R Ratio	206064570174_R02C01.Log R Ratio	206064570174_R02C02.Log R Ratio
IlmnSeq_17:43099552_IlmnFwd	1	0	0.7191634	1.232309	0.6488559	0.7817191
IlmnSeq_17:43099552_IlmnFwd_IlmnDup	1	0	0.8655574	1.179978	0.7319136	0.8444639
rs9701055	1	630053	0.3893069	0.5379636	0.3871996	0.4364024
rs9651229	1	632287	0.4435982	0.9377586	0.5331625	0.6015976
```

Importantly the first three columns must have the names: Name	Chr	Position followed by the samples with suffix either Log.R.Ratio or B.Allele.Freq

**2. Call rates for each sample generated with GenomeStudio:**

Call rates in table format should be generated with GenomeStudio for each sample analyzed in the SNP array. 

Here there is information about how  to generate the input files: https://docs.google.com/document/d/1mXT1G87VnemVMd0mZsD9MbP4VigF6vtQ/edit

Example of call rate table:

```
Index	Sample ID	Call Rate	Gender	p05 Grn	p50 Grn	p95 Grn	p05 Red	p50 Red	p95 Red	p10 GC	p50 GC	Rep Error Rate	PC Error Rate	PPC Error Rate	Aux	Genotype	Score	Subset	Array Info.Sentrix ID	Array Info.Sentrix Position
1	206064570174_R01C01	0,7841907	Unknown	146	1103	6967	423	1383	18108	0,3586062	0,7817292	 	 	 	0		0,00		206064570174	R01C01
2	206064570174_R01C02	0,6686158	Unknown	144	756	6521	417	1123	16369	0,308951	0,7615124	 	 	 	0		0,00		206064570174	R01C02
3	206064570174_R02C01	0,8841922	Unknown	148	1391	7006	464	1684	18739	0,4001076	0,8107659	 	 	 	0		0,00		206064570174	R02C01
4	206064570174_R02C02	0,7895314	Unknown	143	1083	6872	459	1423	18164	0,3614593	0,7856024	 	 	 	0		0,00		206064570174	R02C02
```

Importantly the table must contain the next column names: Index, Sample ID, Call Rate.


**3. Sample sheet in csv format:**
A file that related the sample names with sample IDs used in the SNP array. This file is usually generated by SNP array facilities to be used by GenomeStudio. Format the file to be csv format and compatible with R.
This is an example of file:
```
Sample_ID	SentrixBarcode_A	SentrixPosition_A	SentrixBarcode_B	SentrixPosition_B	Sample_Plate	Sample_Well	Sample_Group	Gender	Sample_Name	Replicate	Parent1	Parent2
ONC.A01_HUS23_ONC_TEGE	2.06E+11	R01C01			HUS_ONCOSYS_2023	A01	not_specified	Unknown	S220_pAdnL_HUS_FFPE		 	 
ONC.C01_HUS23_ONC_TEGE	2.06E+11	R01C02			HUS_ONCOSYS_2023	C01	not_specified	Unknown	S091_pUmbilicus_HUS_a		 	 
ONC.B01_HUS23_ONC_TEGE	2.06E+11	R02C01			HUS_ONCOSYS_2023	B01	not_specified	Unknown	S046_iOme_HUS_FFPE		 	 
ONC.D01_HUS23_ONC_TEGE	2.06E+11	R02C02			HUS_ONCOSYS_2023	D01	not_specified	Unknown	S091_iOme_HUS_b		
```

**4. GCcontent_GSAv3.txt**

File created with this:
https://github.com/VanLoo-lab/ascat/tree/master/LogRcorrection

**5. ReplicationTiming_GSAv3.txt**
